src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala

```diff
@@ -83,6 +86,189 @@ object TaskGroupBuilder extends StrictLogging {
     (executorInfo.build, taskGroup.build, resourceMatch.hostPorts)
   }
 
+  // This method just double-checks that the matched resources are exactly
+  // what is needed and in right quantities.
+  private[this] def verifyMatchedResources(pod: PodDefinition, matchedResources: Seq[mesos.Resource]): Unit = {
+    val cpus = BigDecimal(pod.executorResources.cpus) + pod.containers.map(c => BigDecimal(c.resources.cpus)).sum
+    val mem = BigDecimal(pod.executorResources.mem) + pod.containers.map(c => BigDecimal(c.resources.mem)).sum
+    val disk = BigDecimal(pod.executorResources.disk) + pod.containers.map(c => BigDecimal(c.resources.disk)).sum
+    val gpus = BigDecimal(pod.executorResources.gpus) + pod.containers.map(c => BigDecimal(c.resources.gpus)).sum
+
+    val matchedCpus = matchedResources.filter(_.getName == "cpus").map(c => BigDecimal(c.getScalar.getValue)).sum
+    val matchedMem = matchedResources.filter(_.getName == "mem").map(c => BigDecimal(c.getScalar.getValue)).sum
+    val matchedDisk = matchedResources.filter(_.getName == "disk").map(c => BigDecimal(c.getScalar.getValue)).sum
+    val matchedGpus = matchedResources.filter(_.getName == "gpus").map(c => BigDecimal(c.getScalar.getValue)).sum
+
+    assert(cpus <= matchedCpus)
+    assert(mem <= matchedMem)
+    assert(disk <= matchedDisk)
+    assert(gpus <= matchedGpus)
+  }
+
+  // We have list of (BigDecimal, Resource) pairs in order to avoid using == and != operators with doubles.
+  private case class Resources(
+      cpus: List[(BigDecimal, mesos.Resource)],
+      mem: List[(BigDecimal, mesos.Resource)],
+      disk: List[(BigDecimal, mesos.Resource)],
+      gpus: List[(BigDecimal, mesos.Resource)])
+
+  private class ResourceConsumer(initial: Resources) {
+    private var cpus = initial.cpus
+    private var mem = initial.mem
+    private var disk = initial.disk
+    private var gpus = initial.gpus
+
+    // Find a CPU resource for the given quantity if possible, and remove the resource from the resource list.
+    def consumeCpus(quantity: Double): Option[mesos.Resource] = {
+      consume(cpus, BigDecimal(quantity)).map {
+        case (resources, resource) =>
+          cpus = resources
```

Comment:
This is strange to me. Why are we modifying the state here? How about calling `resource` `consumed_resource`.

Reply:
I have renamed both variables. Why does it look strange to you?